lib/shopify_app/controller_concerns/login_protection.rb

```diff
@@ -5,19 +5,19 @@ module LoginProtection
     class ShopifyDomainNotFound < StandardError; end
 
     included do
+      after_action :set_test_cookie if ShopifyApp.configuration.embedded_app?
       rescue_from ActiveResource::UnauthorizedAccess, :with => :close_session
     end
 
     def shopify_session
-      if shop_session
-        begin
-          ShopifyAPI::Base.activate_session(shop_session)
-          yield
-        ensure
-          ShopifyAPI::Base.clear_session
-        end
-      else
-        redirect_to_login
+      return redirect_to_login unless shop_session
+      clear_top_level_oauth_cookie
```

Comment:
Is this necessary to call on every request?

Maybe it's better to simply ignore the cookie after you no longer need it? Or clear it somewhere during the authentication/cookie-setting process, once.

Reply:
We can't ignore it; we need to remove it or there's a risk that we won't run the full flow the next time we can't recover the session.

We need to wait until we confirm that we've actually recovered the session to remove the cookie, and this seems like the least bad place to do that. We could check that the cookie doesn't exist, but that's probably slower than just deleting the key from the hash regardless.